src/main/java/edu/si/trellis/cassandra/CassandraBinaryService.java

```diff
@@ -0,0 +1,232 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.commons.io.input.BoundedInputStream;
+import org.apache.commons.lang3.Range;
+import org.apache.commons.rdf.api.IRI;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.trellisldp.api.BinaryService;
+import org.trellisldp.api.IdentifierService;
+
+import com.datastax.driver.core.BoundStatement;
+import com.datastax.driver.core.ConsistencyLevel;
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+
+public class CassandraBinaryService implements BinaryService {
+	private static final Logger log = LoggerFactory.getLogger(CassandraBinaryService.class);
+
+	private Session cassandraSession = null;
+
+	private int chunkLength = 1 * 1024 * 1024;
+
+	private static final String SHA = "SHA";
+
+	// TODO JDK9 supports SHA3 algorithms (SHA3_256, SHA3_384, SHA3_512)
+	// TODO Move digest calculation to the C* node.
+	private static final Set<String> algorithms = asList(MD5, MD2, SHA, SHA_1, SHA_256, SHA_384, SHA_512).stream()
+			.collect(toSet());
+
+	private static final String INSERT_QUERY = "INSERT INTO Binarydata (identifier, chunk_index, chunk) VALUES (:identifier, :chunk_index, :chunk)";
+
+	private final PreparedStatement insertStatement;
+
+	private static final String CONTAINS_QUERY = "SELECT identifier FROM Binarydata WHERE identifier = ? and chunk_index = 0;";
+
+	private final PreparedStatement containsStatement;
+
+	private static final String READ_QUERY = "SELECT chunk FROM Binarydata WHERE identifier = ?;";
+
+	private final PreparedStatement readStatement;
+
+	private static final String READ_RANGE_QUERY = "SELECT chunk FROM Binarydata WHERE identifier = ? and chunk_index >= :start and chunk_index <= :end;";
+
+	private final PreparedStatement readRangeStatement;
+
+	private static final String DELETE_QUERY = "DELETE FROM Binarydata WHERE identifier = ?;";
+
+	private final PreparedStatement deleteStatement;
+
+	private IdentifierService idService;
+
+	public CassandraBinaryService(IdentifierService idService, Session session, int chunkLength) {
+		this.idService = idService;
+		this.cassandraSession = session;
+		this.chunkLength = chunkLength;
+		insertStatement = session.prepare(INSERT_QUERY);
+		containsStatement = session.prepare(CONTAINS_QUERY);
+		readStatement = session.prepare(READ_QUERY);
+		readRangeStatement = session.prepare(READ_RANGE_QUERY);
+		deleteStatement = session.prepare(DELETE_QUERY);
+	}
+
+	@Override
+	public Optional<InputStream> getContent(IRI identifier, List<Range<Integer>> ranges) {
+		requireNonNull(ranges, "Byte ranges may not be null");
+		try {
+			if (ranges.isEmpty()) {
+				return Optional.ofNullable(readAll(identifier));
```

Comment:
Does `readAll` ever return `null`? If the binary doesn't exist, I think it just returns an empty `InputStream`...